depends: Set CMAKE_INSTALL_RPATH for native packages

ryanofsky opened the pull request:
This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/projects/10).

---

After #19685 started setting `LDFLAGS`, the `INSTALL_RPATH_USE_LINK_PATH` cmake option used in the libmultiprocess build no longer works, so it is neccessary to set `CMAKE_INSTALL_RPATH` as a fallback.

It's unclear currently whether the bad interaction between `INSTALL_RPATH_USE_LINK_PATH` and `LDFLAGS` is a bug, but the issue is reported:

- https://github.com/bitcoin/bitcoin/issues/19981#issuecomment-696680877
- https://discourse.cmake.org/t/install-rpath-use-link-path-not-working-when-cmake-exe-linker-flags-ldflags-is-set/1892

Fixes #19981

ryanofsky committed:
depends: Set CMAKE_INSTALL_RPATH for native packages

After #19685 started setting LDFLAGS, the INSTALL_RPATH_USE_LINK_PATH cmake
option used in the libmultiprocess build no longer works, so it is neccessary
to set CMAKE_INSTALL_RPATH as a fallback.

It's unclear currently whether the bad interaction between
INSTALL_RPATH_USE_LINK_PATH and LDFLAGS is a bug, but the issue is reported:

  https://github.com/bitcoin/bitcoin/issues/19981#issuecomment-696680877
  https://discourse.cmake.org/t/install-rpath-use-link-path-not-working-when-cmake-exe-linker-flags-ldflags-is-set/1892

Commands useful for building / testing this change

  make -C depends MULTIPROCESS=1 print-libmultiprocess_cmake
  make -C depends MULTIPROCESS=1 print-native_libmultiprocess_cmake
  make -C depends MULTIPROCESS=1 HOST=x86_64-apple-darwin16 print-libmultiprocess_cmake

  rm -rvf depends/x86_64-pc-linux-gnu/native depends/work/staging depends/work/build
  make -C depends MULTIPROCESS=1 V=1 native_libmultiprocess_staged
  for f in `find -name mpgen`; do echo == $f ==; readelf -d $f | grep -i path; done

  make -C depends MULTIPROCESS=1 V=1 native_libmultiprocess_built
  find -name CMakeCache.txt

Fixes #19981

ryanofsky commented:
This commit was originally part of #19160 but I'm making it a standalone PR because it seems to be causing travis timeouts on the macOS 10.12 build that I don't know how to fix:

- https://travis-ci.org/github/bitcoin/bitcoin/jobs/730006940
- https://travis-ci.org/github/bitcoin/bitcoin/jobs/730937293
- https://travis-ci.org/github/bitcoin/bitcoin/jobs/730939841
- https://travis-ci.org/github/bitcoin/bitcoin/jobs/730944459

ryanofsky commented:
Reproduced the travis macOS 10.12 timeout again:

- https://travis-ci.org/github/bitcoin/bitcoin/jobs/731609242

@MarcoFalke do you know any way to debug or fix this? I don't think this PR is actually changing the build results. I think maybe all it is doing is invalidating the travis cache, and that macOS 10.12 travis build is too slow to update the cache. Does this sound right?

MarcoFalke commented:
The build is probably failing altogether. For temporary debugging, you can remove the `&> "/dev/null"`. Though, the `&> "/dev/null"` can't be removed from master, because other builds would fail on travis due to too much output.

ryanofsky commented:
Thanks @MarcoFalke! It still looks to me that this PR might be uncovering a pre-existing problem with the macOS 10.12 build, and not actually causing the failure. Removing >/dev/null in d9dc091b257e9736d67ca8eeecbaf2da6682a3e5 shows the error "configure: error: Qt5Core >= 5.5.1 not found" (https://travis-ci.org/github/bitcoin/bitcoin/jobs/731743605#L2010) with debug output:

```
configure:27263: checking for QT_CORE
configure:27270: $PKG_CONFIG --exists --print-errors "${qt_lib_prefix}Core $qt_version"
Package Qt5Core was not found in the pkg-config search path.
Perhaps you should add the directory containing `Qt5Core.pc'
to the PKG_CONFIG_PATH environment variable
No package 'Qt5Core' found
configure:27273: $? = 1
configure:27287: $PKG_CONFIG --exists --print-errors "${qt_lib_prefix}Core $qt_version"
Package Qt5Core was not found in the pkg-config search path.
Perhaps you should add the directory containing `Qt5Core.pc'
to the PKG_CONFIG_PATH environment variable
No package 'Qt5Core' found
configure:27290: $? = 1
configure:27304: result: no
No package 'Qt5Core' found
configure:27329: error: Qt5Core >= 5.5.1 not found
```

(https://travis-ci.org/github/bitcoin/bitcoin/jobs/731743605#L5266)

It's good to have a starting point for debugging this, but am I wrong that this problem could be happening in master but masked by travis caching?

@dongcarl's PR #19683 also has a failing macOS 10.12 build https://travis-ci.org/github/bitcoin/bitcoin/jobs/730372585, but it seems to be failing with a different problem a little earlier so it's unclear if this Qt5Core error would happen there

dongcarl commented:
FWIW I'm fairly certain the failure in #19683 is unrelated. As it's due to over-specifying `-fuse-ld=` which `clang` treats as a `-Wunused-command-line-argument` warning, but because we also have `-Werror` on, it turns into an error.

ryanofsky commented:
> FWIW I'm fairly certain the failure in #19683 is unrelated. As it's due to over-specifying `-fuse-ld=` which `clang` treats as a `-Wunused-command-line-argument` warning, but because we also have `-Werror` on, it turns into an error.

I'm also sure it's unrelated. But since it happens before my error, if you fix it you might see the same problem there

ryanofsky commented:
It does appear the Qt5Core error is not related to the change in this PR. If I revert the original commit da6c9b7089d2cd0f5380d64882bd18f8794675f1, and just add a comment to funcs.mk 7472ee73d1541d822048c53157adc5030fbef5be, the error still happens: `configure: error: Qt5Core >= 5.5.1 not found`:

- [Diff](https://github.com/ryanofsky/bitcoin/compare/72affcb16cad45bd9e08ca163b2147fd01b84b70..pr/ipc-rpath.3)
- [Branch](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.3)
- https://travis-ci.org/github/bitcoin/bitcoin/jobs/731762168#L2010
- https://travis-ci.org/github/bitcoin/bitcoin/jobs/731762168#L5266-L5281

However, it turns out the Qt5Core error isn't limited to the macOS 10.12 build, but also happens on the CentOS 7 build:

- https://travis-ci.org/github/bitcoin/bitcoin/jobs/731762162#L1512

Which suggests the Qt5Core error is not related to the original macOS 10.12 timeouts https://github.com/bitcoin/bitcoin/pull/20046#issuecomment-701427950 and https://github.com/bitcoin/bitcoin/pull/20046#issuecomment-701473374, but maybe is caused by removing >/dev/null in d9dc091b257e9736d67ca8eeecbaf2da6682a3e5, or maybe is caused by more recent changes to master, or caused by other changes on travis. I guess I will try reverting the >/dev/null change d9dc091b257e9736d67ca8eeecbaf2da6682a3e5 to see if the timeout returns.

ryanofsky commented:
I think I see why the Qt5Core error is happening. I was trying to debug by replacing `source ./ci/test/05_before_script.sh &> "/dev/null` with `source ./ci/test/05_before_script.sh &`, but actually I need to drop the trailing `&` character, otherwise the before_script command doesn't block at all.

ryanofsky commented:
After pushing e5d6dd00509d84de788d9f9f2d5511e2fe7c4c40, was able to reproduce the original macOS 10.12 error and get debug information: https://travis-ci.org/github/bitcoin/bitcoin/jobs/731786312#L4517. The error was `CMake Error: TRY_RUN() invoked in cross-compiling mode` building the `native_cdrkit` package


---


Added 1 commits da6c9b7089d2cd0f5380d64882bd18f8794675f1 -> d9dc091b257e9736d67ca8eeecbaf2da6682a3e5 ([`pr/ipc-rpath.1`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.1) -> [`pr/ipc-rpath.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-rpath.1...pr/ipc-rpath.2)) removing >/dev/null to try to debug macOS 10.12 timeout https://github.com/bitcoin/bitcoin/pull/20046#issuecomment-701473374
Added 2 commits d9dc091b257e9736d67ca8eeecbaf2da6682a3e5 -> 7472ee73d1541d822048c53157adc5030fbef5be ([`pr/ipc-rpath.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.2) -> [`pr/ipc-rpath.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-rpath.2...pr/ipc-rpath.3)) reverting rpath fix to try to debug new Qt5Core error https://github.com/bitcoin/bitcoin/pull/20046#issuecomment-701686770
Added 1 commits 7472ee73d1541d822048c53157adc5030fbef5be -> 9aa445de3f794a92b5e6d0bc21cd854fa2c69619 ([`pr/ipc-rpath.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.3) -> [`pr/ipc-rpath.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-rpath.3...pr/ipc-rpath.4)) adding back >/dev/null to try to reproduce original timeout
Updated 9aa445de3f794a92b5e6d0bc21cd854fa2c69619 -> e5d6dd00509d84de788d9f9f2d5511e2fe7c4c40 ([`pr/ipc-rpath.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.4) -> [`pr/ipc-rpath.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-rpath.4..pr/ipc-rpath.5)) restoring original fix and removing >/dev/null correctly to debug the timeout https://github.com/bitcoin/bitcoin/pull/20046#issuecomment-701723936
Updated e5d6dd00509d84de788d9f9f2d5511e2fe7c4c40 -> 9fe5c425fd70299ee3c43977d14cbee60da4985f ([`pr/ipc-rpath.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.5) -> [`pr/ipc-rpath.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-rpath.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-rpath.5..pr/ipc-rpath.6)) with fix for native_cdrkit error https://travis-ci.org/github/bitcoin/bitcoin/jobs/731786312#L4517

fanquake commented:
ACK 7d0271b5c30f86e4af175a5ab7df5e593cd85195 - I haven't looked in depth, but I've re-read through #19981 and checked the failure by testing #19160 (with this reverted):
```bash
  CXX      bitcoin_node-bitcoind.o
  CXX      init/bitcoin_node-bitcoin-node.o
  GEN      ipc/capnp/echo.capnp.c++
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/native/bin/mpgen: error while loading shared libraries: libcapnpc-0.7.0.so: cannot open shared object file: No such file or directory
make[2]: *** [/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/native/include/mpgen.mk:4: ipc/capnp/echo.capnp.c++] Error 127
```

Hopefully our upstream issue will get a reply at some point. Will wait for Carl to comment.

ryanofsky commented:
Thanks for looking fanquake. Note original version of this PR was more complicated and ugly. Now that it's is just setting CMAKE_INSTALL_RPATH for native packages, I think it's pretty clean and straightforward. It may even a be good thing to set rpath explicitly and be clear the build looks for depends libraries ahead of system libraries

dongcarl commented:
ACK 7d0271b Looked into this a bit, it makes sense that for the things we build in depends, we want the library search to start in depends. It seems reasonable to expect this to happen automatically when `CMAKE_INSTALL_PREFIX` and `INSTALL_RPATH_USE_LINK_PATH` are set, but oh well...

-----

I looked into the CMake codebase a bit, but couldn't find a definite culprit for the weird CMAKE_INSTALL_RPATH_USE_LINK_PATH behaviour. It seems unintended, but we'll see what the forums say I guess.
